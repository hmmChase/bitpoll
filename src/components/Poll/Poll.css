form {
  display: flex;
  flex-direction: column;
  width: 33%;
  height: 300px;
  justify-content: center;
  align-content: center;
  align-items: center;
}

h3 {
  text-align: center;
}

ul {
  padding: 0;
  list-style-type: none;
}

label {
  margin-left: 3px;
}

.submit-btn:disabled {
  background: lightgray;
  color: gray;
}
