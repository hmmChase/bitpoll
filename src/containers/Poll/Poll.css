form {
  display: flex;
  flex-direction: column;
  width: 33%;
  max-width: 300px;
  min-width: 200px;
  justify-content: center;
  align-content: center;
  align-items: center;
  background: #fff;
  text-align: center;
  border: 1px lightblue solid;
  border-radius: 10px;
  margin: 0 auto;
}

.poll-title {
  background: lightsteelblue;
  padding: 10px;
  width: 100%;
  border-radius: 10px;
}

.chart {
  margin: 0 0 0 40px;
}

.tally {
  margin-right: 50px;
  position: relative;
  bottom: 10px;
  margin: 0 auto 10px auto;
}

.tally-left {
  margin-right: 47px;
}

h3 {
  text-align: center;
  margin-bottom: 20px;
}

.submit-btn {
  padding: 10px;
  border-radius: 10px;
  margin-bottom: 20px;
}

.submit-btn:disabled {
  background: lightgray;
  color: gray;
}

.submit-btn:hover {
  background: lightgray;
}

.vote-warning-text {
  padding: 10px;
}

.thanks {
  margin-bottom: 20px;
}

ul {
  padding: 0;
  list-style-type: none;
  margin-bottom: 20px;
}

ul li input[type='radio'] {
  visibility: hidden;
}

ul li label {
  cursor: pointer;
  margin-left: 5px;
  font-size: 1.5em;
}

ul li .check {
  display: inline-block;
  border: 5px solid #aaaaaa;
  border-radius: 100%;
  height: 25px;
  width: 25px;
  transition: border 0.25s linear;
  -webkit-transition: border 0.25s linear;
}

ul li .check::before {
  transition: background 0.25s linear;
  -webkit-transition: background 0.25s linear;
}

input[type='radio']:checked ~ .check {
  border: 10px solid #77e29c;
}

input[type='radio']:checked ~ .check::before {
  background: #77e29c;
}
