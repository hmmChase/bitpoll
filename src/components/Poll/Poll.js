import React, { Component } from 'react';
import './Poll.css';

class Poll extends Component {
  constructor(props) {
    super(props);

    this.state = {
      answer: null,
      disabled: true
    };
  }

  componentDidMount = () => {};

  handleChange = event => {
    this.setState({
      answer: event.target.value,
      disabled: false
    });
  };

  handleSubmit = event => {
    event.preventDefault();

    alert(this.state.answer);
  };

  render() {
    return (
      <form onSubmit={this.handleSubmit}>
        <p>Should the blocksize be increased?</p>

        <ul>
          <li>
            <input
              className="option"
              id="option1"
              type="radio"
              value="yes"
              checked={this.state.answer === 'yes'}
              onChange={this.handleChange}
            />
            <label htmlFor="option1">Yes</label>
          </li>

          <li>
            <input
              className="option"
              id="option2"
              type="radio"
              value="no"
              checked={this.state.answer === 'no'}
              onChange={this.handleChange}
            />
            <label htmlFor="option2">No</label>
          </li>
        </ul>

        <button
          className="submit-btn"
          type="submit"
          disabled={this.state.disabled}
        >
          Vote
        </button>
      </form>
    );
  }
}

export default Poll;
