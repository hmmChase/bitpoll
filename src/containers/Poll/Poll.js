import React, { Component } from 'react';
import { connect } from 'react-redux';
import firebase from '../../utils/firebase';
import * as actions from '../../actions';
import Options from './Options/Options';
import Results from './Results/Results';
import './Poll.css';

export class Poll extends Component {
  constructor(props) {
    super(props);

    this.state = {
      option1: 0,
      option2: 0
    };
  }

  componentDidMount = () => {
    this.syncDB();
  };

  ifVoted = voterIDs => {
    const hasVoted = voterIDs.includes(this.props.userId);
    this.props.storeIfVoted(hasVoted);
  };

  syncDB = () => {
    const pollRef = firebase
      .database()
      .ref()
      .child('poll');
    const option1Ref = pollRef.child('option1');
    const option2Ref = pollRef.child('option2');
    const voterIDsRef = pollRef.child('voterIDs');
    option1Ref.on('value', snapshot => {
      const option1RefVal = snapshot.val();
      this.props.storeOption1Tally(option1RefVal);
      this.setState({
        option1: option1RefVal
      });
    });
    option2Ref.on('value', snapshot => {
      const option2RefVal = snapshot.val();
      this.props.storeOption2Tally(option2RefVal);
      this.setState({
        option2: option2RefVal
      });
    });
    voterIDsRef.on('value', snapshot => {
      const voterIDsRef = snapshot.val();
      const voterIDs = voterIDsRef ? [this.state.voterIDs, ...voterIDsRef] : [];
      this.ifVoted(voterIDs);
    });
  };

  handleSubmit = event => {
    event.preventDefault();
    this.increment();
  };

  increment = () => {
    let optionValue = this.state[this.props.selectedOption];
    const incrementOption = optionValue + 1;
    this.updateResult(incrementOption);
  };

  updateResult = incrementOption => {
    this.setState({
      [this.props.selectedOption]: incrementOption
    });
    this.updateDB(incrementOption);
  };

  updateDB = incrementOption => {
    const pollRef = firebase
      .database()
      .ref()
      .child('poll');
    pollRef.update({
      [this.props.selectedOption]: incrementOption,
      voterIDs: [this.props.userId]
    });
  };

  voteButton = () => (
    <button
      className="submit-btn"
      type="submit"
      disabled={this.props.voteBtnDisabled}
    >
      Vote
    </button>
  );

  render() {
    return (
      <form onSubmit={this.handleSubmit}>
        <h3 className="poll-title">Should the blocksize be increased?</h3>
        {this.props.userId && this.props.ifVoted ? (
          <div>
            <Results />
            <p className="thanks">Thanks for voting!</p>
          </div>
        ) : this.props.userId && this.props.isContributor ? (
          <div>
            <Options />
            <div>{this.voteButton()}</div>
          </div>
        ) : (
          <p className="vote-warning-text">
            You must be signed in and have contributed to the Bitcoin repo in
            order to vote.
          </p>
        )}
      </form>
    );
  }
}

export const mapStateToProps = state => ({
  userId: state.user.userId,
  isContributor: state.user.isContributor,
  ifVoted: state.user.ifVoted,
  selectedOption: state.poll.selectedOption,
  value: state.poll.value,
  voteBtnDisabled: state.poll.voteBtnDisabled
});

export const mapDispatchToProps = dispatch => ({
  storeIfVoted: hasVoted => dispatch(actions.storeIfVoted(hasVoted)),
  storeOption1Tally: tally => dispatch(actions.storeOption1Tally(tally)),
  storeOption2Tally: tally => dispatch(actions.storeOption2Tally(tally))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Poll);
