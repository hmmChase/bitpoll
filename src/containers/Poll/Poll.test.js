import React from 'react';
import { shallow } from 'enzyme';
import Poll from './Poll';

describe('Poll', () => {
  let poll;

  beforeEach(() => {
    poll = shallow(<Poll />);
  });

  it('matches the snapshot', () => {
    expect(poll).toMatchSnapshot();
  });
});
