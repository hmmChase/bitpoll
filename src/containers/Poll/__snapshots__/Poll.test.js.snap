// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Poll matches the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Poll />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h3
          className="poll-title"
>
          Should the blocksize be increased?
</h3>,
        <p
          className="vote-warning-text"
>
          You must be signed in and have contributed to the Bitcoin repo in order to vote.
</p>,
      ],
      "onSubmit": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Should the blocksize be increased?",
          "className": "poll-title",
        },
        "ref": null,
        "rendered": "Should the blocksize be increased?",
        "type": "h3",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "You must be signed in and have contributed to the Bitcoin repo in order to vote.",
          "className": "vote-warning-text",
        },
        "ref": null,
        "rendered": "You must be signed in and have contributed to the Bitcoin repo in order to vote.",
        "type": "p",
      },
    ],
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h3
            className="poll-title"
>
            Should the blocksize be increased?
</h3>,
          <p
            className="vote-warning-text"
>
            You must be signed in and have contributed to the Bitcoin repo in order to vote.
</p>,
        ],
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Should the blocksize be increased?",
            "className": "poll-title",
          },
          "ref": null,
          "rendered": "Should the blocksize be increased?",
          "type": "h3",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "You must be signed in and have contributed to the Bitcoin repo in order to vote.",
            "className": "vote-warning-text",
          },
          "ref": null,
          "rendered": "You must be signed in and have contributed to the Bitcoin repo in order to vote.",
          "type": "p",
        },
      ],
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
